### PR TITLE
[BEAM-970] add side-input support to gearpump-runner

### PR DESCRIPTION
**Pull request by manuzhang**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by manuzhang** [comment]
R: @kennknowles

**Comment by kennknowles** [comment]
Jenkins failure is the same dependency problem that affected many PRs. I will review as green.

File: runners/gearpump/pom.xml [hunk 1]
```diff
@@ -70,26 +70,17 @@
...
-                      org.apache.beam.sdk.transforms.CountTest,
-                      org.apache.beam.sdk.transforms.FlattenTest,
-                      org.apache.beam.sdk.transforms.ParDoTest,
-                      org.apache.beam.sdk.transforms.SampleTest,
-                      org.apache.beam.sdk.transforms.ViewTest,
-                      org.apache.beam.sdk.transforms.join.CoGroupByKeyTest
+                      org.apache.beam.sdk.transforms.FlattenTest
                     </exclude>
-                    <!-- merging windows is not supported in Gearpump -->
+                    <!-- assertion should have failed-->
```
Comment (kennknowles) [inline runners/gearpump/pom.xml]: Which assertion?

File: runners/gearpump/pom.xml [hunk 2]
```diff
@@ -70,26 +70,17 @@
...
                     <exclude>
-                      org.apache.beam.sdk.transforms.windowing.WindowingTest,
-                      org.apache.beam.sdk.util.ReshuffleTest
+                      org.apache.beam.sdk.testing.PAssertTest
+                    </exclude>
+                    <!-- only one side input is supported -->
```
Comment (kennknowles) [inline runners/gearpump/pom.xml]: Most of the logic seems to support multiple side inputs - using a union value and writing to a `SideInputHandler`. If really only one side input is supported, I think we need to use a new JUnit `@Category` to exclude only the tests that use more than one side input. If we exclude `ParDoTest` we lose too much core functionality, including a lot of side input testing.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/functions/DoFnFunction.java [hunk 3]
```diff
@@ -18,77 +18,144 @@
...
 import org.apache.beam.runners.core.DoFnRunners;
+import org.apache.beam.runners.core.PushbackSideInputDoFnRunner;
+import org.apache.beam.runners.core.SideInputHandler;
 import org.apache.beam.runners.gearpump.GearpumpPipelineOptions;
 import org.apache.beam.runners.gearpump.translators.utils.DoFnRunnerFactory;
 import org.apache.beam.runners.gearpump.translators.utils.NoOpAggregatorFactory;
 import org.apache.beam.runners.gearpump.translators.utils.NoOpStepContext;
 import org.apache.beam.sdk.transforms.DoFn;
-import org.apache.beam.sdk.util.SideInputReader;
+import org.apache.beam.sdk.transforms.join.RawUnionValue;
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/functions/DoFnFunction.java]: I think that we shouldn't have a dependency on this. Either make a new class that does the same thing here, or we need to promote this type to the `values` namespace. I think making a new class just for here is probably better.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/utils/ParDoTranslatorUtils.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  public static <InputT> JavaStream<RawUnionValue> withSideInputStream(
+      TranslationContext context,
+      JavaStream<WindowedValue<InputT>> inputStream,
+      Collection<PCollectionView<?>> sideInputs) {
+    JavaStream<RawUnionValue> mainStream =
+        inputStream.map(new ToRawUnionValue<InputT>(0), "map_to_RawUnionValue");
+
+    final int size = sideInputs.size();
+    if (size > 1) {
+      throw new UnsupportedOperationException("only one side input is supported");
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/utils/ParDoTranslatorUtils.java]: Why do you need to reject this? It seems like if you support one side input via union values + `SideInputHandler`, then you can support as many as you need.

**Comment by manuzhang** [comment]
@kennknowles  First of all, thanks for taking a look during your holidays 👍. 
For the tests,
1. I was under the **wrong** impression that three-way merge hadn't been supported in Gearpump and rejected multiple side inputs (embarrassing...)
2. `PAssertTest#testContainsInAnyOrderFalse` and `PAssertTest#testEmptyFalse` don't pass because of `assertion should have failed`. I haven't figured out why.
3. I lazily copied `RawUnionValue` to gearpump-runner. It's also used in flink-runner so it could be valuable to move `RawUnionValule` to `values`.

**Comment by kennknowles** [comment]
Do the gearpump local tests take too long for Jenkins? One thing that helped a lot for Apex was to have them running in precommit while things were under development. You could also add them to precommit or to Travis in one commit just for the PR and then I can remove before the merge. Obviously it cannot be on all the time, but right now sharing the failure stack traces, etc, would be a good way to collaborate.

**Comment by kennknowles** [comment]
I expect that the problem with `PAssert` is that the `AssertionFailure` is not being propagated.

**Comment by kennknowles** [comment]
Just noting that this PR has acquired merge conflicts.

**Comment by manuzhang** [comment]
Thanks, I will rebase

**Comment by manuzhang** [comment]
superseded by https://github.com/apache/beam/pull/2150

**Pull request closed without merging**